README.md
# XChemExplorer-Py3
based on xchem / XChemExplorer
An attempt to update XChemExplorer to Python 3, for use on non-Diamond computers


ChangeLog.txt
v1.9.4 (10/10/2022)
- catch unusual space group names in PDB CRYST card

v1.9.3 (29/06/2022)
- added 'module load ccp4/7.1.018' to grade script otherwise the helper script will run with python3

v1.9.2 (28/06/2022)
- replace obsolete 'source /dls/science/groups/i04-1/software/pandda-update/ccp4/ccp4-7.0/bin/ccp4.setup-sh' with 'module load ccp4/7.1.016' before running pandda.export


v1.9.1 (16/06/2022)
- changed add instances of 'module load ccp4' to 'module load ccp4/7.1.018', except for the pandda2 function

v1.9.0 (22/04/2022)
- add functionality for running PanDDA2 at Diamond Light Source
- fixed hyperlink to PanDDA manual

v1.8.2 (20/01/2022)
- added 'Porphyromonas gingivalis' to NCBI taxonomy id
- fixed SARS-CoV-2 name in NCBI taxonomy id table
- bug fix: XChemDeposit - added missing underscore to loop statement in _pdbx_audit_support block
- XChemDeposit: issue warning and do not add empty xtal string to error list
- bug fix: XChemDeposit - recognition of staraniso logfile

v1.8.1 (19/01/2022)
- fix: event map to sf conversion; ignore existing ligand pdb files

v1.8.0 (17/12/2021)
- added feature for working with cocktails/ combi-soaks: multiple compounds/ smiles separated by semi-colon in db are recognised during restraint generation

v1.7.36 (13/10/2021)
- bug fix: manual selection of different auto-processing results with same pipeline
- removed obsolete paths used for parsing of auto-processing results

v1.7.35 (13/10/2021)
- ignore json merging statistics and xia2.mmcif if aimless.log exists

v1.7.34 (25/06/2021)
- added pdbx_audit_support category for funding agency now required for PDB deposition

v1.7.33 (02/02/2021)
- bug fix: enable RefinementParameters button in XChemCOOT

v1.7.32 (24/01/2021)
- bug fix: XChemDeposit - enable reading of not symlinked logfiles

v1.7.31 (20/01/2021)
- bug fix: XChemDeposit - enable reading of staraniso .table1 file

v1.7.30 (17/12/2020)
- bug fix in XChemDeposit

v1.7.29 (09/12/2020)
- bug fix: syntax error in XChemDeposit
- XChemDeposit: use default pdb_extract if no installation in XChemExplorer_dir

v1.7.28 (25/11/2020)
- added exception to catch 'Cannot allocate memory' error in XChemMain.get_gda_barcodes

v1.7.27 (24/11/2020)
- bug fix: add missing DataProcessingUniqueReflectionsLow/High column to db

v1.7.26 (12/11/2020)
- XChemDeposit: remove \n and \r from smiles string in _diffrn.details

v1.7.25 (11/11/2020)
- convert json logfile from dials into pseudo aimless file for mmcif file preparation (required for pdb-extract)

v1.7.24 (02/11/2020)
- capitalized organism names in NCBI_taxonomy_ID dictionary

v1.7.23 (02/11/2020)
- pandda refinement: stopped sourcing obsolete pandda v0.2.12 at DLS (otherwise refinement update script fails due to missing gemmi library)

v1.7.22 (28/10/2020)
- cosmetic changes to deposition interface
- bug fix: make event map assignment during MMCIF preparation backward compatible
- bug fix: finding of event map with highest CC for given ligand

v1.7.21 (19/10/2020)
- bug fix: use default CCP4 for giant.merge_conformations

v1.7.20 (16/10/2020)
- bug fix: remove undocumented and faulty changes to RunQuickRefine

v1.7.19 (12/10/2020)
- bug fix: use 'module load ccp4/7.0.072' for pandda.analyse, but 'module load ccp4' (i.e. v7.1) for pre-run ground-state selection at DLS

v1.7.18 (12/10/2020)
- bug fix: added missing 'module load ccp4' statement when generating pandda.analyse shell script at DLS

v1.7.17 (09/10/2020)
- bug fix: corrected positions of ignore checkboxes in pandda.analyse

v1.7.16 (08/10/2020)
- enable export of selected datasets with pandda.export
- glitch fix: remove hydrogens before saving model for buster refinement (causes problems during ligand occupancy refinement)
- XChemRefine: removed "-q medium.q" option for non-DLS refinement
- added helper script to which reverts to last successful non-pandda refinement
- XChemCoot (pandda refinement): reading of (2)fofc.maps disabled because Coot 0.9 does not handle P1 maps well
- pandda.analyse: stopped sourcing obsolete pandda v0.2.12 at DLS

v1.7.15 (25/09/2020)
- added option to ignore GELLY sanity check in BUSTER
- added option to ignore GELLY sanity check during pandda export with BUSTER

v1.7.14 (24/09/2020)
- added option for ligand occupancy refinement in BUSTER
- XChemCootBuster: reading of (2)fofc.maps disabled because Coot 0.9 does not handle P1 maps well

v1.7.13 (23/09/2020)
- bug fix: use wavelength read from (.free).mtz during model MMCIF preparation when updating SF MMCIF

v1.7.12 (14/09/2020)
- bug fix: problem with recognizing AIMLESS logfiles during preparation of MMCIF files
- bug fix: try finding experimental wavelength from .free.mtz file if it is 0.0 in refine.mtz during MMCIF file preparation
- preparation of mmcif files: catch and report missing event mtz files

v1.7.11 (09/09/2020)
- bug fix: preparation of mmcif files: run aimless in onlymerge mode if only json file available

v1.7.10 (08/09/2020)
- bug fix - preparation of mmcif files for deposition: enable usage of event mtz files which were generated during 'BUSTER' export routine

v1.7.9 (03/09/2020)
- change 'Open COOT' to 'Open COOT - REFMAC refinement -' in Refinement action box

v1.7.8 (01/09/2020)
- bug fix: string formatting error during restraints generation on local machine
- added scanchirals option to rhofit command in fit_ligands function
- bug fix: added 'module load buster' to fit_ligands function

v1.7.7 (22/07/2020)
- bug fix: logical error in creation of restraints script when adding 'module load ccp4/phenix' command
- removed phenix.reduce step from restraints generation

v1.7.6 (22/07/2020)
- bug fix: added 'module load ccp4/phenix' to restraints script in case not part of user bashrc file

v1.7.5 (21/07/2020)
- bug fix: remove obsolete 'module load mx' statement from xce_<restraints_prg> script which currently loads a gemmi incompatible ccp4 version

v1.7.4 (21/07/2020)

lib/XChemDialogs.py
import sys
from PyQt4 import QtGui, QtCore

import XChemDB

class select_columns_to_show(QtGui.QDialog):
    def __init__(self, data_source_file,parent = None):
        super(select_columns_to_show, self).__init__(parent)
        self.columns_in_data_source=XChemDB.data_source(data_source_file).return_column_list()

        self.column_dict={}

        layout = QtGui.QVBoxLayout(self)
        number_of_entries=len(self.columns_in_data_source)
        columns_shown_in_dialog_column=25
        grid = QtGui.QGridLayout()
        x=0
        y=0
        columns_to_ignore=['Sample ID','ID']
        for entries_added in range(number_of_entries):
            if not self.columns_in_data_source[entries_added][1] in columns_to_ignore:
                data_source_column = QtGui.QCheckBox(self.columns_in_data_source[entries_added][1])
                self.column_dict[entries_added]=data_source_column
#            data_source_column.toggle()
                grid.addWidget(data_source_column, y,x)
                y+=1
            if y==columns_shown_in_dialog_column:
                y=0
                x+=1
        layout.addLayout(grid)


        # OK and Cancel buttons
        buttons = QtGui.QDialogButtonBox(
            QtGui.QDialogButtonBox.Ok | QtGui.QDialogButtonBox.Cancel,
            QtCore.Qt.Horizontal, self)
        buttons.accepted.connect(self.accept)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)

    def data_source_column_active(self):
        columns_to_show=['Sample ID']
        for key in self.column_dict:
            if self.column_dict[key].isChecked():
                columns_to_show.append(self.columns_in_data_source[key][1])
        return columns_to_show

    # static method to create the dialog and return (date, time, accepted)
    @staticmethod
    def return_selected_columns(parent = None):
        dialog = select_columns_to_show(parent)
        result = dialog.exec_()
        columns = dialog.data_source_column_active()
        return (columns,result == QtGui.QDialog.Accepted)


web/jscss/js/create_view.js
function create_view(div_name,pdb_bound,event_name,lig_name) {
// Create NGL Stage object
var stage = new NGL.Stage(div_name);
// Handle window resizing
window.addEventListener( "resize", function( event ){
    stage.handleResize();
}, false );
// Code for example: test/map-shift
Promise.all( [
    //stage.loadFile( "data://"+event_name+".ccp4" ),
    //stage.loadFile( "data://"+event_name+".pdb" )
    stage.loadFile( event_name ),
    stage.loadFile( pdb_bound ),

] ).then( function( ol ){
    var map = ol[ 0 ];
    var struc = ol[ 1 ];
    struc.autoView(lig_name);
    var surfRepr = map.addRepresentation( "surface", {
        boxSize: 10,
        useWorker: false,
        wrap: true,
        color: "skyblue",
        contour: true
    } );
    struc.addRepresentation( "cartoon" );
    struc.addRepresentation( "ball+stick", { sele: "hetero" } );
    stage.setFocus( 95 );
    stage.mouseObserver.signals.scrolled.add( function( delta ){
        if( stage.mouseObserver.altKey ){
            var d = Math.sign( delta ) / 5;
            var l = surfRepr.getParameters().isolevel;
            surfRepr.setParameters( { isolevel: l + d } );
        }
    } );
} );
}



helpers/assign_stereochemistry.py
import os,sys
import glob
from rdkit import Chem
#from rdkit.Chem.EnumerateStereoisomers import EnumerateStereoisomers, StereoEnumerationOptions

sys.path.append(os.path.join(os.getenv('XChemExplorer_DIR'),'lib'))
import XChemDB

def enumerateStereoChem(compoundID,sampleDir,db,xtal):
    # first need to recreate the original CIF file with phenix.elbow because we cannot be sure
    # which program was used to create the initial restrains
    # this is because funny things happen to aromatic rings in case the file was made with GRADE
    os.chdir(os.path.join(sampleDir,'compound'))
    sql = "select CompoundSMILES from mainTable where CrystalName = '%s'" % xtal
    print 'sqlite query:',sql
    query = db.execute_statement(sql)
    originalSMILES = query[0][0]
    print originalSMILES
    cmd = 'phenix.elbow --smiles="%s" --id=LIG --output=tmp' % (originalSMILES)
    print 'current working directory:',os.getcwd()
    print 'runnning:',cmd
    os.system(cmd)

    stereosmiles = None
    pdb = None
    if os.path.isfile(os.path.join(sampleDir,'compound','tmp.pdb')):
        pdb = os.path.join(sampleDir,'compound','tmp.pdb')
    else:
        print 'cannot find tmp.pdb'
        pass
    if pdb:
        mol = Chem.MolFromPDBFile(pdb)
        Chem.AssignStereochemistry(mol,cleanIt=True,force=True,flagPossibleStereoCenters=True)
        if Chem.FindMolChiralCenters(mol,includeUnassigned=True) == []:
            print 'no chiral centres found'
            db_dict = {}
            db_dict['CompoundStereo'] = 'FALSE'
            updateDB(db,db_dict,xtal)
        else:
            stereosmiles = Chem.MolToSmiles(mol,isomericSmiles=True)
            generateRestraints(compoundID,sampleDir,db,stereosmiles,xtal)

def generateRestraints(compoundID,sampleDir,db,stereosmiles,xtal):
    cmd = 'phenix.elbow --smiles="%s" --chiral=enumerate --id=LIG --output=%s' %(stereosmiles,compoundID)
    os.system(cmd)
    checkFiles(compoundID,sampleDir,db,stereosmiles,xtal)


def checkFiles(compoundID,sampleDir,db,stereosmiles,xtal):
    foundCIF = False
    allCIF = ''
    for cif in glob.glob(os.path.join(compoundID + '_*.cif')):
        foundCIF = True
        allCIF += cif+';'
        print '---',cif,'===',allCIF
    if foundCIF:
        db_dict = {}
        db_dict['CompoundStereo'] = 'TRUE'
        db_dict['CompoundStereoSMILES'] = stereosmiles
        db_dict['CompoundStereoCIFprogram'] = 'phenix.elbow'
        db_dict['CompoundStereoCIFs'] = allCIF[:-1]
        print '>>>',db_dict
        updateDB(db,db_dict,xtal)

def updateDB(db,db_dict,xtal):
    # update stereo field
    # update restraintsprogram field
    # update stereosmiles
    db.update_data_source(xtal, db_dict)

if __name__=='__main__':
    compoundID = sys.argv[1]
    sampleDir = sys.argv[2]
    xtal = sampleDir[sampleDir.rfind('/')+1:]
    print 'sampleID:',xtal
    database = sys.argv[3]
    print 'database:',database
    db=XChemDB.data_source(database)
    enumerateStereoChem(compoundID,sampleDir,db,xtal)



helpers/find_best_fitting_ligand.py
import sys, os
import glob

sys.path.append(os.path.join(os.getenv('XChemExplorer_DIR'),'lib'))
import XChemDB

def parse_autofit_folder(sampleDir,compoundID):
    filenames = ['best.pdb','ligand_fit_1.pdb']
    pdbList = []
    for files in glob.glob(os.path.join(sampleDir,'autofit_ligand',compoundID+'_*','*')):
        pdb = files[files.rfind('/')+1:]
        if pdb in filenames:
            pdbList.append(files)
    return pdbList

def find_input_mtz(sampleDir):
    possibleMTZ = ['refine.mtz','init.mtz','dimple.mtz',]
    mtzin = None
    for mtz in possibleMTZ:
        if os.path.isfile(os.path.join(sampleDir,mtz)):
            mtzin = os.path.join(sampleDir,mtz)
            break
    return mtzin

def calculate_cc(sampleDir,compoundID,db):
    xtal = sampleDir[sampleDir.rfind('/') + 1:]
    pdbList = parse_autofit_folder(sampleDir,compoundID)
    mtzin = find_input_mtz(sampleDir)
    for pdb in pdbList:
        autofitDir = pdb[:pdb.rfind('/')]
        os.chdir(autofitDir)
        os.system('phenix.get_cc_mtz_pdb %s %s > phenix.get_cc_mtz_pdb.log' %(pdb,mtzin))
    new_file_links(pdbList)
    bestRun, bestCC = parse_cc_log(sampleDir, compoundID)
    changeLinks(sampleDir, pdbList, bestRun, compoundID)
    fitting_program = None
    if 'phenix' in bestRun:
        fitting_program = 'phenix.ligandfit'
    elif 'rhofit' in bestRun:
        fitting_program = 'rhofit'
    db_dict = {}
    db_dict['CompoundAutofitprogram'] = fitting_program
    db_dict['CompoundAutofitCC'] = str(bestCC)
    updateDB(db, db_dict, xtal)


def new_file_links(pdbList):
    for pdb in pdbList:
        autofitDir = pdb[:pdb.rfind('/')]
        autofitRun = pdb.split('/')[len(pdb.split('/')) - 2]
        pdbFile = pdb[pdb.rfind('/')+1:]
        os.chdir(autofitDir)
        if not os.path.isfile(autofitRun+'.pdb'):
            os.system('ln -s %s %s.pdb' %(pdbFile,autofitRun))

def changeLinks(sampleDir,pdbList,bestRun,compoundID):
    os.chdir(sampleDir)
    for pdb in pdbList:
        autofitDir = pdb.split('/')[len(pdb.split('/')) - 2]
        if autofitDir == bestRun:
            if os.path.isfile(pdb.replace(sampleDir,'.')):
                os.system('/bin/rm %s.pdb' %compoundID)
                os.system('ln -s %s %s.pdb' %(pdb.replace(sampleDir,'.'),compoundID))

def parse_cc_log(sampleDir,compoundID):
    ccDict = {}
    ccList = []
    for ccLog in glob.glob(os.path.join(sampleDir,'autofit_ligand',compoundID+'_*','phenix.get_cc_mtz_pdb.log')):
        autofitDir = ccLog.split('/')[len(ccLog.split('/'))-2]
        for line in open(ccLog):
            if line.startswith('local CC:') and len(line.split()) > 2:
                cc = line.split()[2]
                ccList.append(float(cc))
                ccDict[autofitDir] = cc
                break
    bestRun = None
    try:
        bestCC = max(ccList)
        for ccRun in ccDict:
            print ccRun,ccDict[ccRun],bestCC
            if str(ccDict[ccRun]) == str(bestCC):
                bestRun = ccRun
                break
    except ValueError:
        pass
    return bestRun, bestCC


def updateDB(db,db_dict,xtal):
    db.update_data_source(xtal, db_dict)


if __name__ == '__main__':
    compoundID = sys.argv[1]
    sampleDir = sys.argv[2]
    database = sys.argv[3]
    db=XChemDB.data_source(database)

    calculate_cc(sampleDir, compoundID, db)

helpers/resort_ligand_atoms.py
# last edited: 07/06/2017, 17:00

import os,sys
sys.path.append(os.path.join(os.getenv('XChemExplorer_DIR'),'lib'))

def get_atom_order_of_ensemble_model(ensembleModel):

    ligandResnames = ['LIG','FRS','DRG','UNL']

    ensembleLIGdir = {}

    for line in open(ensembleModel):
        if line.startswith('ATOM') or line.startswith('HETATM'):
            atomName_line=  str(line[11:16])
            altLoc_line=    str(line[16:17])
            resname_line=   str(line[17:20])
            chainID_line=   str(line[20:23])
            resseq_line=    str(line[23:26])

            residueID = resname_line.replace(' ','')+'-'+\
                        chainID_line.replace(' ','')+'-'+\
                        resseq_line.replace(' ','')+'-'+\
                        chainID_line.replace(' ','')+'-'+\
                        altLoc_line

            if resname_line in ligandResnames:
                if residueID not in ensembleLIGdir:
                    ensembleLIGdir[residueID]=[]
                ensembleLIGdir[residueID].append([atomName_line,altLoc_line,resname_line,chainID_line,resseq_line])

    return ensembleLIGdir

def resort_ligand_atoms_in_refined_model(refinedModel,ensembleLIGdir):
    ligandResnames = ['LIG','FRS','DRG','UNL']
    out=''
    refineLIGDir={}
    LIGlineAdditional=''
    LIGlist=[]
    for line in open(refinedModel):
        if line.startswith('ATOM') or line.startswith('HETATM'):
            altLoc_line=    str(line[16:17])
            resname_line=   str(line[17:20])
            chainID_line=   str(line[20:23])
            resseq_line=    str(line[23:26])

            residueID = resname_line.replace(' ','')+'-'+\
                        chainID_line.replace(' ','')+'-'+\
                        resseq_line.replace(' ','')+'-'+\
                        chainID_line.replace(' ','')+'-'+\
                        altLoc_line

            if residueID in ensembleLIGdir:
                if residueID not in refineLIGDir:
                    refineLIGDir[residueID]=[]
                else:
                    # if a ligand was for modelled after pandda.inspect
                    LIGlineAdditional+=line
                refineLIGDir[residueID].append(line)


    LIGline=''
    for ligand in ensembleLIGdir:
        for atomLine in ensembleLIGdir[ligand]:
            atomName=atomLine[0]
            for line in refineLIGDir[ligand]:
                atomName_line=str(line[11:16])
                if atomName_line==atom:
                    LIGline+=line
                    break

    for line in open(refinedModel):
        if line.startswith('ATOM') or line.startswith('HETATM'):
            resname_line=   str(line[17:20])
            if resname_line in ligandResnames:
                continue
            else:
                out+=line

        elif line.startswith('END'):
            continue

        elif line.startswith('TER'):
            continue

        else:
            out+=line

    out+=LIGline
    out+='END\n'

    os.system('/bin/mv %s %s' %(refinedModel,refinedModel.replace('.pdb','_original_atom_order.pdb')))
    f=open(refinedModel,'w')
    f.write(out)
    f.close()



if __name__=='__main__':
    ensembleModel=sys.argv[1]
    refinedModel=sys.argv[2]
    ensembleLIGdir=get_atom_order_of_ensemble_model(ensembleModel)
    resort_ligand_atoms_in_refined_model(refinedModel,ensembleLIGdir)